op-e2e/sgt/sgt_test.go

```diff
@@ -0,0 +1,389 @@
+			txValue:         big.NewInt(10001),
+			expectedErr:     errorInsufficientBalance,
+			action:          fullSGTGasPaymentAndNonZeroTxValueWithInsufficientNativeBalanceFail,
+		},
+		{
+			name:            "PartialSGTGasPaymentAndNonZeroTxValueWithInsufficientNativeBalanceFail",
+			depositSgtValue: big.NewInt(10000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(10000000000000 - 10000),
+			expectedErr:     errorInsufficientBalance,
+			action:          partialSGTGasPaymentAndNonZeroTxValueWithInsufficientNativeBalanceFail,
+		},
+	}
+
+	for index, tCase := range tests {
+		t.Run(tCase.name, func(t *testing.T) {
+			tCase.action(t, ctx, int64(index), tCase.depositSgtValue, tCase.depositL2Value, tCase.txValue, sgt)
+		})
+	}
+}
+
+func setUpTestAccount(t *testing.T, ctx context.Context, index int64, sgt *SgtHelper, depositSgtValue *big.Int, depositL2Value *big.Int) (*ecdsa.PrivateKey, common.Address) {
+	opts := &bind.CallOpts{Context: ctx}
+	rng := rand.New(rand.NewSource(index))
+	testPrivKey := testutils.InsecureRandomKey(rng)
+	testAddr := crypto.PubkeyToAddress(testPrivKey.PublicKey)
+
+	// check it's a fresh account
+	sgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, int64(0), sgtBalance.Int64())
+	l2Balance, err := sgt.L2Client.BalanceAt(ctx, testAddr, nil)
+	require.NoError(t, err)
+	require.Equal(t, int64(0), l2Balance.Int64())
+
+	// deposit initial sgt and native(L2) balance to the test account
+	sgt.depositSgtAndNativeFromGenesisAccountToAccount(t, ctx, testAddr, depositSgtValue, depositL2Value)
+	// ensure that sgt and native balance of testAccount are correctly initialized
+	preSgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, depositSgtValue.Cmp(preSgtBalance), 0)
+	preL2Balance, err := sgt.L2Client.BalanceAt(ctx, testAddr, nil)
+	require.NoError(t, err)
+	require.Equal(t, depositL2Value.Cmp(preL2Balance), 0)
+	return testPrivKey, testAddr
+}
+
+// balance invariant check: preTotalBalance = postTotalBalance + gasCost + txValue
+func invariantBalanceCheck(t *testing.T, ctx context.Context, sgt *SgtHelper, addr common.Address, gasCost *big.Int, txValue *big.Int, preSgtBalance *big.Int, preL2Balance *big.Int, postSgtBalance *big.Int) {
+	postL2Balance, err := sgt.L2Client.BalanceAt(ctx, addr, nil)
+	require.NoError(t, err)
+	preBalance := preSgtBalance.Add(preSgtBalance, preL2Balance)
+	postBalance := postSgtBalance.Add(postSgtBalance, gasCost)
+	postBalance = postBalance.Add(postBalance, txValue)
+	postBalance = postBalance.Add(postBalance, postL2Balance)
+	require.Equal(t, 0, preBalance.Cmp(postBalance))
+}
+
+func nativaGasPaymentWithoutSGTSuccess(t *testing.T, ctx context.Context, index int64, sgtValue *big.Int, l2Value *big.Int, txValue *big.Int, sgt *SgtHelper) {
+	// ensure that sgt balance is 0
+	require.Equal(t, common.Big0.Cmp(sgtValue), 0)
```

Comment:
Instead of asserting the parameters, why not initialize the parameters right in this function?

Reply:
Good point! Currently, I'm a bit uncomfortable switching between the test pattern of defining inputs and verifying outputs in OP and our test pattern of verifying logic with assertions on the fly...

Done in https://github.com/ethstorage/optimism/pull/125/commits/8aafd47a87f33baff6be561ec8820430adb01033